### DISCUSSION TITLE
Do you have ETAs for support of JS app event hooks and Computed fields

### DISCUSSION THREAD

**Discussion by titomitto** [pocketbase/pocketbase#1831] [Q&A] [answered]
@ganigeorgiev do you have estimations for when features like JS app event hooks and computed fields might arrive?
I have a project where I use deno for writting services and I had some issues with subscriptions since EventSource is not supported in the backend, app event hooks might come in handy. I would also love the ability to create stats from specific tables which could be solved with the computed fields/RawSQL feature

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure yet, it is difficult to say.

JS app event hooks were initially scheduled for v0.13/v0.14 but later decided to postpone it since I'm planning a slight refactoring of the hooks structure in the near future.

Option to allow creating aggregations, and any custom `SELECT` query in general, will be available in the upcoming v0.13 release.

> I had some issues with subscriptions since EventSource is not supported in the backend

I'm not sure what env you are targetting but for node or react native you can load a polyfill - https://github.com/pocketbase/js-sdk#nodejs-via-npm.

**Reply by denisgurev** [reply]
@ganigeorgiev By "aggregations" do you mean that there will be support for sum/avg?

**Reply by ganigeorgiev** [reply]
@denisgurev Yes, but not only those. There will be more info once I finalize the implementation, but the general idea is to allow users to specify any `SELECT` query (including joins, subqueries, CTE, etc. statements) and convert that into a collection, preserving the same access controls (list and view), filter and sort capabilities as seamless as possible.

For example, the following imaginary query:

```sql
SELECT
  collectionA.id,
  collectionC.id as cId,
  collectionB.file,
  SUM(collectionB.total) as total
FROM collectionA
LEFT JOIN collectionB on collectionB.a = collectionA.id
LEFT JOIN collectionC on collectionC.id = collectionA.c
WHERE ... # some condition
GROUP BY (collectionA.id)
```

will result to a read-only collection with the following fields:
- id (pk)
- cId (`relation` to `collectionC`)
- file (`file`; note that the actual files are not copied and the access to them is proxied)
- total (`number`)

The only requirement to the query will be to have a unique `id` column (_this is to preserve the same `getOne("RECORD_ID")` api; if the query doesn't have a suitable unique `id` column, an autogenerated could be used, eg. `(ROW_NUMBER() OVER(ORDER BY 1)) as id`_)

I've already created a local POC, but there are still some edge cases that need to be handled and test properly before making them public - allowing relations between views, referencing file fields across multiple views (eg. view->view->*->base), special handling for cascade delete on reverse and/or circular relations (eg. base->view->base), etc. Once I finish the tests and the required UI updates, will push the changes in the `develop` branch.

**Reply by titomitto** [reply]
> I'm not sure what env you are targetting but for node or react native you can load a polyfill - https://github.com/pocketbase/js-sdk#nodejs-via-npm.

I am currently writing all my backend services in Deno. The npm module works well in deno except for the realtime subscriptions which require EventSource to work.

**Reply by ganigeorgiev** [reply]
@TitoMitto It should be the same for Deno. You can either try to load one of the the npm polyfills as shown in the link or use a Deno-only library. I'm not much familiar with the Deno ecosystem, but you could check for example https://deno.land/x/eventsource@v0.0.3. The main idea is to load it into the `global/globalThis` context so that the SDK can use it, something like (_I haven't tested it_):

```js
import { EventSource } from "https://deno.land/x/eventsource@v0.0.3/mod.ts";

globalThis.EventSource = EventSource;
```
